src/order/conditionally_complete_lattice.lean

```diff
@@ -483,6 +483,29 @@ end
 protected lemma Inf_le {s : set ℕ} {m : ℕ} (hm : m ∈ s) : Inf s ≤ m :=
 by { rw [nat.Inf_def ⟨m, hm⟩], exact nat.find_min' ⟨m, hm⟩ hm }
 
+lemma nonempty_of_pos_Inf {s : set ℕ} (h : 0 < Inf s) : s.nonempty :=
+begin
+  by_contradiction contra, rw set.not_nonempty_iff_eq_empty at contra,
+  have h' : Inf s ≠ 0, { exact ne_of_gt h, }, apply h',
+  rw nat.Inf_eq_zero, right, assumption,
+end
+
+lemma Inf_interval_above_eq_succ_iff {s : set ℕ}
```

Comment:
This is the only lemma where I'm not sure if this is the mathlibbest way to do things. The rest of the PR looks good to me.
@PatrickMassot @urkud can you take a look at this please?

Reply:
Thanks, for the review and agreed: this lemma is where I felt least sure about what was best.

Reply:
I just had a look at how I would prove this lemma and I wrote the following. I'm not sure it's better, but at least it extracts two more lemmas that don't seem to be there already. That `nat.has_Inf` is pretty specific, using both that `nat` is bounded below and the fact it's a discrete order, I'm not sure we should spend too much time generalizing this.
```lean
lemma nat.eq_Ici_of_nonempty_of_upward_closed {s : set ℕ} (hs : s.nonempty)
  (hs' : ∀ (k₁ k₂ : ℕ), k₁ ≤ k₂ → k₁ ∈ s → k₂ ∈ s) : s = Ici (Inf s) :=
ext (λ n, ⟨λ H, nat.Inf_le H, λ H, hs' (Inf s) n H (nat.Inf_mem hs)⟩)

lemma nat.nonempty_of_Inf_eq_succ {s : set ℕ} {k : ℕ} (h : Inf s = k + 1) : s.nonempty :=
nat.nonempty_of_pos_Inf (h.symm ▸ (nat.succ_pos k) : Inf s > 0)

lemma Inf_interval_above_eq_succ_iff {s : set ℕ}
  (hs : ∀ (k₁ k₂ : ℕ), k₁ ≤ k₂ → k₁ ∈ s → k₂ ∈ s) (k : ℕ) :
  Inf s = k + 1 ↔ k + 1 ∈ s ∧ k ∉ s :=
begin
  split,
  { intro H,
    rw show s = Ici (Inf s),
      from nat.eq_Ici_of_nonempty_of_upward_closed (nat.nonempty_of_Inf_eq_succ H) hs,
    simp [H] },
  { rintro ⟨H, H'⟩,
    rw [nat.Inf_def (⟨_, H⟩ : s.nonempty), nat.find_eq_iff],
    exact ⟨H, λ n hnk hns, H' $ hs n k (nat.lt_succ_iff.mp hnk) hns⟩ } 
end
```

Reply:
Thank you very much @PatrickMassot. I have just pushed (a slightly-modified version of) your code above.

I also agree that the two new lemmas are worth having in their own right.

Reply:
@PatrickMassot thanks a lot!